GH-40074: [C++][FS][Azure] Implement `DeleteFile()` for flat-namespace storage accounts

felipecrv opened the pull request:
### Rationale for this change

It was not implemented yet.

### What changes are included in this PR?

 - An implementation of `DeleteFile()` that is specialized to storage accounts that don't have HNS support enabled
 - This fixes a semantic issue: deleting a file should not delete the parent directory when the file deleted was the last one
 - Increased test coverage
 - Fix of a bug in the version that deletes files in HNS-enabled accounts (we shouldn't let `DeleteFile` delete directories even if they are empty)

### Are these changes tested?

Yes. Tests were re-written and moved to `TestAzureFileSystemOnAllScenarios`.
* Closes: #40074

felipecrv commented:
@av8or1

felipecrv commented:
@Tom-Newton

kou commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -2055,6 +2035,116 @@ class AzureFileSystem::Impl {
+  /// \pre location.path does not end with a trailing slash.
+  Status DeleteFileOnFileSystem(const DataLake::DataLakeFileSystemClient& adlfs_client,
+                                const AzureLocation& location, bool require_file_to_exist,
+                                Azure::Nullable<std::string> lease_id = {}) {
```
When do we need `lease_id`?

felipecrv replied:
This was needed for an experiment I did with the API before, but I'm removing this together with the implementation of `MovePaths` for flat-namespace accounts.

felipecrv replied:
I removed it in the latest commits.

kou commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -2055,6 +2035,116 @@ class AzureFileSystem::Impl {
+      options.AccessConditions.LeaseId = file_blob_lease_guard->LeaseId();
+    }
+    try {
+      auto response = blob_client.Delete(options);
```
Do we need to call `file_blob_lease_guard->BreakBeforeDeletion()` before `Delete()`?

felipecrv replied:
This was surprising to me as well. In the Azure Blobs API if I delete a blob that’s holding a lease I can release the lease on that path just fine. Re-creating a Blob on the same path immediately after deletion+release works just fine. This was a mess in the DataLake API, but is fine in the Blobs API.

kou replied:
Wow. Could you add the information as a comment of here or `LeaseGuard::BreakBeforeDeletion()`?

felipecrv replied:
Alright. Adding a comment on `BreakBeforeDeletion`.

kou commented on cpp/src/arrow/filesystem/azurefs_test.cc:
```diff
@@ -889,6 +889,117 @@ class TestAzureFileSystem : public ::testing::Test {
+        fs()->DeleteFile(data.Path("nonexistent-file")));
+    EXPECT_RAISES_WITH_MESSAGE_THAT(
+        IOError,
+        ::testing::HasSubstr("Not a regular file: '" + data.Path("nonexistent-file/") +
+                             "'"),
+        fs()->DeleteFile(data.Path("nonexistent-file/")));
```
```suggestion
        ::testing::HasSubstr("Not a regular file: '" + data.Path("nonexistent-dir/") +
                             "'"),
        fs()->DeleteFile(data.Path("nonexistent-dir/")));
```

felipecrv replied:
This is supposed to be just like the test above, but with a slash in the path. I will replace "file" with "path" so I can keep them the same.

kou commented on cpp/src/arrow/filesystem/azurefs_test.cc:
```diff
@@ -889,6 +889,117 @@ class TestAzureFileSystem : public ::testing::Test {
+    EXPECT_RAISES_WITH_MESSAGE_THAT(
+        IOError,
+        ::testing::HasSubstr("Not a regular file: '" +
+                             data.Path("dir/nonexistent-file/") + "'"),
+        fs()->DeleteFile(data.Path("dir/nonexistent-file/")));
```
```suggestion
                             data.Path("dir/nonexistent-dir/") + "'"),
        fs()->DeleteFile(data.Path("dir/nonexistent-dir/")));
```

felipecrv replied:
Same here.

kou commented on cpp/src/arrow/filesystem/azurefs_test.cc:
```diff
@@ -889,6 +889,117 @@ class TestAzureFileSystem : public ::testing::Test {
+    // Recreating the file on the same path gurantees leases were properly released/broken
+    setup_dir_file0();
+
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir/file0"), FileType::File);
+    EXPECT_RAISES_WITH_MESSAGE_THAT(
+        IOError,
+        ::testing::HasSubstr("Not a regular file: '" + data.Path("dir/file0/") + "'"),
+        fs()->DeleteFile(data.Path("dir/file0/")));
```
How about swapping these assertions to check `DeleteFile("dir/file0")` while `dir/file0` exists?

```suggestion
    EXPECT_RAISES_WITH_MESSAGE_THAT(
        IOError,
        ::testing::HasSubstr("Not a regular file: '" + data.Path("dir/file0/") + "'"),
        fs()->DeleteFile(data.Path("dir/file0/")));
    arrow::fs::AssertFileInfo(fs(), data.Path("dir/file0"), FileType::File);
```

felipecrv replied:
Done.

kou commented on cpp/src/arrow/filesystem/azurefs_test.cc:
```diff
@@ -889,6 +889,117 @@ class TestAzureFileSystem : public ::testing::Test {
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir/"), FileType::Directory);
+    arrow::fs::AssertFileInfo(fs(), data.Path("dir/file0"), FileType::NotFound);
+
+    // Recreating the file on the same path gurantees leases were properly released/broken
```
How about using different directory  instead for it? It'll be cleaner.

felipecrv replied:
But I'm using the same path to test the lease management code.

kou commented on cpp/src/arrow/filesystem/azurefs_test.cc:
```diff
@@ -1528,6 +1639,21 @@ TYPED_TEST(TestAzureFileSystemOnAllScenarios, DeleteDirContentsFailureNonexisten
 }
 
+TYPED_TEST(TestAzureFileSystemOnAllScenarios, DeleteFileAtRoot) {
+  void TestDeleteFileAtRoot();
```
```suggestion
  this->TestDeleteFileAtRoot();
```

felipecrv commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -2544,7 +2635,28 @@ Status AzureFileSystem::DeleteRootDirContents() {
 Status AzureFileSystem::DeleteFile(const std::string& path) {
   ARROW_ASSIGN_OR_RAISE(auto location, AzureLocation::FromString(path));
-  return impl_->DeleteFile(location);
+  if (location.container.empty()) {
+    return Status::Invalid("DeleteFile requires a non-empty path.");
+  }
+  if (internal::HasTrailingSlash(location.path) || location.path.empty()) {
+    // Container paths (locations w/o path) or paths ending with a slash are not file
+    // paths.
+    return NotAFile(location);
+  }
```
It's not the same logic I'm implementing here and it obscures the pre-conditions I'm guaranteeing before delegating to `DeleteFileOnFileSystem/DeleteFileOnContainer`.

kou replied:
Hmm. It seems that both of `ValidateFileLocation()` and here check `location.container.empty()`, `location.path.empty()` and `HasTrailingSlash()`.

felipecrv replied:
Same checks, but different errors returned. Anyways, I changed the logic even more now and the trailing slashes are handled inside `DeleteFileOnFileSystem/DeleteFileOnContainer` instead.

felipecrv commented:
@kou I added a commit making error messages a bit more detailed now and removed the `WithHierarchicalNamespace()` check splitting the tests because the different version were behaving differently in some situations.

kou commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -2055,6 +2057,148 @@ class AzureFileSystem::Impl {
+    DCHECK(!location.path.empty());
+    constexpr auto kFileBlobLeaseTime = std::chrono::seconds{15};
+    auto no_trailing_slash_location = location.RemoveTrailingSlash(
+        /*preserve_original=*/true);
```
Hmm. I think that `preserve_original` isn't a good approach. It will confuse us. Is it really needed?

felipecrv replied:
I want `all` to be path the user used so error messages make sense.

felipecrv replied:
What about `RemoveTrailingSlashFromPath()`?

kou replied:
It seems that `no_trailing_slash_location.all` isn't used by the user:

```cpp
      ARROW_ASSIGN_OR_RAISE(auto file_info,
                            GetFileInfo(container_client, no_trailing_slash_location));
```

`GetFileInfo()` in the above code may return an error to the user. But `GetFileInfo()` doesn't use `no_trailing_slash_location.all` for an error message:

https://github.com/apache/arrow/blob/a03d957b5b8d0425f9d5b6c98b6ee1efa56a1248/cpp/src/arrow/filesystem/azurefs.cc#L1275-L1277

It uses `file_client.GetUrl()` and it's based on `no_trailing_slash_location.path` not `.all`.

And it seems that `FileInfo::path()` of `file_info` returned by `GetFileInfo()` isn't used in this lambda. (`file_info.path()` uses `no_trailing_slash_location.all` but it's not used.)

felipecrv replied:
Alright. I'm fixing `GetFileInfo`, then.

kou commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -2055,6 +2039,132 @@ class AzureFileSystem::Impl {
+      }
+    }
+
+    DCHECK(!internal::HasTrailingSlash(location.path));
```
Can we remove this because we have `if (internal::HasTrailingSlash(location.path)) {` above?

felipecrv replied:
Removing.

kou commented on cpp/src/arrow/filesystem/azurefs.cc:
```diff
@@ -2055,6 +2039,132 @@ class AzureFileSystem::Impl {
+  /// \pre location.path is not empty.
+  Status DeleteFileOnContainer(const Blobs::BlobContainerClient& container_client,
+                               const AzureLocation& location, bool require_file_to_exist,
+                               const char* operation) {
```
It seems that we don't need to receive `operation` as an argument. How about defining it as a local variable instead of an argument?

felipecrv replied:
I might use `Move()` with it, so I will keep it.

kou reviewed (approved):
+1
